### ISSUE TITLE
@context breaks the picking up of the @id

### ISSUE THREAD

**Issue by AlasdairGray** [jathanasiou/hw-dissertation-validata2#4]
When a code block includes `@context` then it stops the code picking up the available resource identifiers. For example the following minor addition to the sample code block stops it from working
```
{
  "@context": "http://schema.org/",
  "@id":"http://njh.me/genid-c7a50d6e828f4595a195f9c67304f6d2-0AE87CB066C8D9CD43117E0C5A3090F0",
  "@type":["https://schema.org/Dataset"],
  "https://schema.org/description": "The GeneChip® Drosophila Genome Array is a microarray tool for studying expression of Drosophila melanogaster transcripts.",
  "https://schema.org/identifier": "http://www.affymetrix.com/products/arrays/specific/fly.affx",
  "https://schema.org/name": "Affymetrix array: GeneChip Drosophila Genome 2.0 Array",
  "https://schema.org/keywords": "gene tool",
  "https://schema.org/url": [{"@id":"https://www.flymine.org/flymine/dataset"}]
}
```

**Comment by jathanasiou** [comment]
@AlasdairGray this JSON-LD snipped seems to validate fine for me so I can't replicate the issue. Could you give it a try again?  

One possibility is that the RDFShape service was updated and fixed that issue, if it was occurring on that layer.

Another possibility is that it was occurring during the JSON-LD parsing and due to [this library update](https://github.com/jathanasiou/hw-dissertation-validata2/commit/a71783f15958e072d3a7d28c3a975808e80eacd4#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R9) on `jsonld` and `n3` it no longer occurs in latest builds.

**Comment by AlasdairGray** [comment]
I've updated to the latest codebase locally and still have the problem. Just looked in the console and saw the following CORS errors
```
Access to XMLHttpRequest at 'http://schema.org/' from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
xhr.js:115 GET http://schema.org/ net::ERR_FAILED
```
Same issue on the deployed version

**Comment by jathanasiou** [comment]
I was trying to investigate the issue but had trouble operating the app in localhost.

I also tried the deployed Validata2 and it seems that the RDFShape server is not responding right now, their client is also having problems. Deploying a HW-operated server might be useful for stability at this point.

**Comment by AlasdairGray** [comment]
The issue causing this problem manifests prior to making the call to the RDFShape server. An error appears when data containing a schema.org in the @context field.

**Comment by jathanasiou** [comment]
So, I have investigated the issue a bit more.

# Replication
[This error](https://github.com/jathanasiou/hw-dissertation-validata2/issues/4#issuecomment-619986852) occurs while _entering_ a snippet with `context` due to CORS (schema.org and application domain differ). Browser CORS-bypassing plugins need to be turned off to replicate this.

# Cause
N3 is used to parse the RDF triples and when having a `context` property it, evidently, tries to access the context domain, which throws the CORS exception.

# Impact
Due to occurring during parsing, this prevents the Node list from being populated, putting the tool in an unusable state:

<details>
<summary>Screenshot:</summary>

![Failed State](https://user-images.githubusercontent.com/8679925/81590066-c4e01f80-93a9-11ea-85ac-a2dbc10d57f1.png)
</details>

# Resolution
Will have to prevent N3 from attempting such a request, if it offers such an option, or catch the exception and fail silently. This might still impact parsed RDF nodes.

**Comment by jathanasiou** [comment]
@AlasdairGray  as a temporary workaround, using an addon like [Allow CORS](https://chrome.google.com/webstore/detail/allow-cors-access-control/lhobafahddgcelffkeicbaginigeejlf) should prevent this issue for a user.

I will consider solutions in the meanwhile.

**Comment by AlasdairGray** [comment]
Using a browser extension is not an option. We cannot rely on users having or being able/willing to configure such an extension.

**Comment by AlasdairGray** [comment]
I have found a workaround using a [Custom Document Loader](https://github.com/digitalbazaar/jsonld.js/blob/b6862c711c286547c0cbf442b83d0f3bb8f8fd0a/README.md#custom-document-loader). This allows us to perform the schema.org redirect from the declared context to the location that the actual file is located.

**Comment by jathanasiou** [comment]
The custom loader solution does seem to work as expected. I hadn't noticed that  the `jsonld` package itself had support for that.

In my local tests I noticed that

```
  "@context": "http://schema.org/",
```

was still failing due to the trailing `/` in the URL. I did try removing that in the loader but there were still exceptions in later processing by `jsonld`. Is this expected behavior?

**Comment by AlasdairGray** [comment]
> In my local tests I noticed that
> `  "@context": "http://schema.org/",`
> was still failing due to the trailing / in the URL. 

I'll need to add that permutation to the set of URLs (easily done).

> I did try removing that in the loader but there were still exceptions in later processing by jsonld. Is this expected behavior?

Did you get the popup error saying that there was a problem with the RDFShape or was it another error? Are you doing this on the server deployment (which this solution still does not work on) or from your local machined?

**Comment by AlasdairGray** [comment]
Although today it may be retrieving the context document but very slowly.

**Comment by AlasdairGray** [comment]
Fixed in 1623545
